Make the explanation for variant ordering clearer (#43)
wheelnext/wheelnext: docs/proposals/pepxxx_wheel_variant_support.md

#### Variant Ordering

The order of variant properties is defined first by the order of namespaces, then the order of features in the namespace
and then the order of feature values in a feature.

The order of namespaces is defined by `default-priorities.namespace`. The order of features in a namespace is defined by
their order in the variant provider output, but can be overridden by `default-priorities.feature`. Features in

> `default-priorities.feature` are sorted before features from the variant provider output. Similarly, the order of

Reviewer: Yeah, I never was really sure which explanation is easier to comprehend. It's either:
1. Take values from plugin, then move these from TOML first; or
2. Take values from TOML, then add missing values from the plugin.

Perhaps indeed 2. is easier. I mean, you can just say "iterate in order, if it's missing, add to the end", which is easier than "moving to front".

> feature values in a feature is defined by their order in the order in the variant provider output,

Reviewer: Maybe "property values"?

`default-priorities.property` which take precedence.

> A variant wheel has a higher priority than another variant wheel if its most important property is more important than

Reviewer: Yeah, it's quite a mouthful. And perhaps would use a sentence of introduction, something akin "Variants wheels are sorted according to the properties they were built with."

the most important other variant wheel. If both wheels have the same most important property, compare the second most
important property for each, and so on, until a tie-breaker is found.

> A different way to describe the same algorithm: Assign each namespace, feature name in a namespace and feature value in

Reviewer: This is a bit too abstract for me to judge if it's correct. The "according to their sorted translated property lists" doesn't seem to really say *how* they are sorted.

#### Variant Ordering

To determine which variant wheel to install when multiple wheels are compatible, variant wheels are ordered by their
variant properties.

The order of variant properties is defined first by the order of namespaces, then the order of features in the namespace
and then the order of property values in a feature.

> The order of namespaces is defined by `default-priorities.namespace`. The order of features in a namespace is defined by

Reviewer: ```suggestion
The order of namespaces is defined by `default-priorities.namespace`. The order of features in a namespace is initially defined by
```

> `default-priorities.feature`. Features not listed in `default-priorities.feature` are appended from their order in the
> variant provider output. Similarly, the order of property values in a feature is defined by their order in

Reviewer: ```suggestion
`default-priorities.feature`. Features not listed in `default-priorities.feature` are appended in the order they are
returned by the variant provider plugin. Similarly, the order of property values in a feature is defined by their order in
```

`default-priorities.property`, missing properties are appended from their order in the variant provider output.

A variant wheel has a higher priority than another variant wheel if its most important property is more important than
the most important other variant wheel. If both wheels have the same most important property, compare the second most
important property for each, and so on, until a tie-breaker is found.

> A different way to describe the same algorithm: Use the index in `default-priorities.namespace` to assign each namespace

Reviewer: I'm afraid this is very hard to follow, even when you know the algorithm. Perhaps an actual pseudo-code would work better. Something like:

```python
namespaces = metadata["default-priorities.namespace"]
for namespace in namespaces:
  features[namespace] = metadata.get(f"default-priorities.feature.{namespace}", []).append(providers[namespace].features)
  for feature in features[namespace]:
    properties[namespace][feature] = metadata.get(f"default-priorities.property.{namespace}.{feature}", []).append(providers[namespace].properties[feature])

for property in variant_properties:
  property_score = (namespaces.index(property.namespace, ...)
```

Sorry, this is as far as I can go without actually testing. That said, I still have no clue if the algorithm you're proposing is actually correct. I see we've switched variantlib to reodering lists over using score tuples, and there was probably a reason we did that. There were some corner cases that the algorithm failed at previously.

https://github.com/wheelnext/variantlib/blob/main/variantlib/resolver/sorting.py

Author: What I'm trying to describe is https://github.com/astral-sh/uv/blob/829ffa1cbb2f6db2358017e9e40d9b37e7c5b27d/crates/uv-variants/src/resolved_variants.rs#L55-L104. The logic is complex, that's why I added both explanations, I think it's fine if people understand either of them. We can reference the variantlib implementation in the reference implementation section.

Reviewer: I don't know yet, that's the point. I'm sorry but given how hard this part is, I can't just sign off on it unless I know that it works, and I can't know if it works unless I'm able to make an implementation out of it.

Also, for the sake of simplification we can skip the "not in" part, appending duplicate features/values to the list won't change the end result.

#### Variant Ordering

To determine which variant wheel to install when multiple wheels are compatible, variant wheels are ordered by their
variant properties.

The order of variant properties is defined first by the order of namespaces, then the order of features in the namespace
and then the order of property values in a feature.

The order of namespaces is defined by `default-priorities.namespace`. The order of features in a namespace is initially defined by
`default-priorities.feature`. Features not listed in `default-priorities.feature` are appended in the order they are
returned by the variant provider plugin. Similarly, the order of property values in a feature is defined by their order in

> `default-priorities.property`, missing properties are appended from their order in the variant provider output.

Author: I tried to keep that part short, it should be evident if you got it the wrong way round.
